File: TypeScript/PaymentsService/src/Infrastructure/Payments/GooglePay/GooglePayPaymentClient.ts
import * as uuid from 'uuid';
import { IPaymentClient } from "../../../Domain/Payments/IPaymentClient";
import { ProcessPaymentBaseRequest } from "../../../Domain/Payments/ProcessPaymentBaseRequest";
import { ProcessPaymentBaseResponse } from "../../../Domain/Payments/ProcessPaymentBaseResponse";
import { ProcessGooglePayPaymentRequest } from "./ProcessGooglePayPaymentRequest";
import { ProcessGooglePayPaymentResponse } from "./ProcessGooglePayPaymentResponse";

export class GooglePayPaymentClient implements IPaymentClient {
    processPayment(request: ProcessPaymentBaseRequest): ProcessPaymentBaseResponse {
        const req = request as ProcessGooglePayPaymentRequest;

        if (!req) {
            throw new Error("Invalid argument");
        }

        const resp: ProcessGooglePayPaymentResponse = new ProcessGooglePayPaymentResponse();
        resp.email = req.email;
        resp.transactionId = uuid.v4();

        if (req.token === "fail") {
            resp.status = "Failed";
            resp.error = "Generic error";
        } else {
            resp.status = "Success";
        }

        return resp;
    }
}

File: CSharp/PaymentsService/Infrastructure/Payments/GooglePay/GooglePayPaymentClient.cs
using PaymentsService.Domain.Payments;

namespace PaymentsService.Infrastructure.Payments.GooglePay;

public sealed class GooglePayPaymentClient : IPaymentClient
{
    public ProcessPaymentBaseResponse ProcessPayment(ProcessPaymentBaseRequest request)
    {
        var req = request as ProcessGooglePayPaymentRequest ?? throw new ArgumentException();

        ProcessGooglePayPaymentResponse resp = new()
        {
            Email = req.Email,
            TransactionId = Guid.NewGuid().ToString()
        };

        if (req.Token == "fail")
        {
            resp.Status = "Failed";
            resp.Error = "Generic error";
        }
        else
        {
            resp.Status = "Success";
        }

        return resp;
    }
}
